app/src/main/java/org/openobservatory/ooniprobe/activity/oonirun/OoniRunV2Activity.kt

```diff
@@ -0,0 +1,217 @@
+
+
+        // If the intent contains a link, but the `link_id` is zero.
+        if (possibleRunId == 0L) {
+            finishWithError().run { return }
+        }
+        val executor = TaskExecutor()
+        binding.cancelButton.setOnClickListener {
+            executor.cancelTask()
+            finishWithError(message = getString(R.string.Modal_Cancel))
+        }
+        executor.executeTask({
+            try {
+                return@executeTask descriptorManager.fetchDescriptorFromRunId(
+                    possibleRunId,
+                    this
+                )
+            } catch (exception: Exception) {
+                exception.printStackTrace()
+                ThirdPartyServices.logException(exception)
+                return@executeTask null
+            }
+        }) { descriptorResponse: TestDescriptor? ->
+            fetchDescriptorComplete(
+                descriptorResponse
+            )
+        }
+    }
+
+    /**
+     * Shows an error message and ends the activity.
+     *
+     * @param message The error message to be shown.
+     */
+    private fun finishWithError(message: String = getString(R.string.Modal_Error)) {
+        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
+        finish()
+    }
+
+
+    /**
+     * The task to fetch the descriptor from the link is completed.
+     *
+     *
+     * This method is called when the `fetchDescriptorFromRunId` task is completed.
+     * The `descriptorResponse` is the result of the task.
+     * If the task is successful, the `descriptorResponse` is the descriptor.
+     * Otherwise, the `descriptorResponse` is null.
+     *
+     *
+     * If the `descriptorResponse` is not null, start the `AddDescriptorActivity`.
+     * Otherwise, show an error message.
+     *
+     * @param descriptorResponse The result of the task.
+     * @return null.
+     */
+    private fun fetchDescriptorComplete(descriptorResponse: TestDescriptor?) {
+        descriptorResponse?.let {
+            startActivity(AddDescriptorActivity.newIntent(this, descriptorResponse))
+        } ?: run {
+            Toast.makeText(this, getString(R.string.Modal_Error), Toast.LENGTH_LONG).show()
+        }
+    }
+
+    /**
+     * Extracts the run id from the provided Uri.
+     * The run id can be in two different formats:
+     *
+     *
+     * 1. ooni://runv2/link_id
+     * 2. https://run.test.ooni.org/v2/link_id
+     *
+     *
+     * The run id is the `link_id` in the link.
+     * If the Uri contains a link, but the `link_id` is not a number, null is returned.
+     * If the Uri contains a link, but it is not a supported link, null is returned.
+     *
+     *
+     *
+     * @param uri The Uri data.
+     * @return The run id if the Uri contains a link with a valid `link_id`, or null otherwise.
+     */
+    private fun getRunId(uri: Uri): Long? {
+        val host = uri.host
+        try {
+            when (host) {
+                "runv2" -> {
+                    /**
```

Comment:
```suggestion
                    /*
```

doesn't need to be JavaDoc